probe/client/postgres/postgres.go

```diff
@@ -97,60 +98,75 @@ func (r *PostgreSQL) checkData() error {
-			db := sql.OpenDB(pgdriver.NewConnector(clientOptions...))
-			if db == nil {
-				return false, "OpenDB error"
-			}
-			// query the data
-			log.Debugf("[%s / %s / %s] - SQL - [%s]", r.ProbeKind, r.ProbeName, r.ProbeTag, sqlstr)
-			rows, err := db.Query(sqlstr)
-			if err != nil {
-				return false, fmt.Sprintf("Query error - [%s], %v", v, err)
-			}
-			if !rows.Next() {
-				rows.Close()
-				return false, fmt.Sprintf("No data found for [%s]", k)
-			}
-			//check the value is equal to the value in data
-			var value string
-			if err := rows.Scan(&value); err != nil {
-				rows.Close()
-				return false, err.Error()
-			}
-			if value != v {
-				rows.Close()
-				return false, fmt.Sprintf("Value not match for [%s] expected [%s] got [%s] ", k, v, value)
-			}
-			rows.Close()
-			db.Close()
-			log.Debugf("[%s / %s / %s] - Data Verified Successfully! - [%s] : [%s]", r.ProbeKind, r.ProbeName, r.ProbeTag, k, v)
+		return r.ProbeWithDataChecking()
+	}
+	return r.ProbeWithPing()
+}
+
+// ProbeWithPing do the health check with ping & Select 1;
+func (r *PostgreSQL) ProbeWithPing() (bool, string) {
+	r.ClientOptions = append(r.ClientOptions, pgdriver.WithDatabase("template1"))
+	db := sql.OpenDB(pgdriver.NewConnector(r.ClientOptions...))
+	if db == nil {
+		return false, "OpenDB error"
+	}
+	defer db.Close()
+
+	if err := db.Ping(); err != nil {
+		return false, err.Error()
+	}
+
+	// run a SQL to test
+	row, err := db.Query(`SELECT 1`)
+	if err != nil {
+		return false, err.Error()
+	}
+	row.Close()
+	return true, "Check PostgreSQL Server Successfully!"
+}
+
+// ProbeWithDataChecking do the health check with data checking
+func (r *PostgreSQL) ProbeWithDataChecking() (bool, string) {
+	if len(r.Data) == 0 {
+		log.Warnf("[%s / %s / %s] - No data found, use ping instead", r.ProbeKind, r.ProbeName, r.ProbeTag)
+		return r.ProbeWithPing()
+	}
+
+	for k, v := range r.Data {
+		log.Debugf("[%s / %s / %s] - Verifying Data - [%s] : [%s]", r.ProbeKind, r.ProbeName, r.ProbeTag, k, v)
+		//connect to the database
+		dbName, sqlstr, err := r.getSQL(k)
+		if err != nil {
+			return false, fmt.Sprintf("Invalid SQL data - [%s], %v", v, err)
 		}
-	} else {
-		r.ClientOptions = append(r.ClientOptions, pgdriver.WithDatabase("template1"))
-		db := sql.OpenDB(pgdriver.NewConnector(r.ClientOptions...))
+		clientOptions := append(r.ClientOptions, pgdriver.WithDatabase(dbName))
+		db := sql.OpenDB(pgdriver.NewConnector(clientOptions...))
 		if db == nil {
 			return false, "OpenDB error"
 		}
-		defer db.Close()
-
-		if err := db.Ping(); err != nil {
-			return false, err.Error()
-		}
-
-		// run a SQL to test
-		row, err := db.Query(`SELECT 1`)
+		// query the data
+		log.Debugf("[%s / %s / %s] - SQL - [%s]", r.ProbeKind, r.ProbeName, r.ProbeTag, sqlstr)
+		rows, err := db.Query(sqlstr)
 		if err != nil {
+			return false, fmt.Sprintf("Query error - [%s], %v", v, err)
+		}
+		if !rows.Next() {
+			rows.Close()
```

Comment:
need call `db.Close()` in error cases?

Reply:
nice catch!

Reply:
I put the code in for loop into function`verifyData()`, and use the `defer db.Close() and defer rows.Close()` inside. 

Please review it again